packages/vite/src/node/plugins/dynamicImportVars.ts

```diff
@@ -0,0 +1,218 @@
+    return null
+  }
+  const { globParams, rawPattern, userPattern } = dynamicImportPattern
+  const params = globParams
+    ? `, ${JSON.stringify({ ...globParams, import: '*' })}`
+    : ''
+  const exp = `(import.meta.glob(${JSON.stringify(userPattern)}${params}))`
+
+  return {
+    rawPattern,
+    pattern: userPattern,
+    glob: exp
+  }
+}
+
+export function dynamicImportVarsPlugin(config: ResolvedConfig): Plugin {
+  const resolve = config.createResolver({
+    preferRelative: true,
+    tryIndex: false,
+    extensions: []
+  })
+  const { include, exclude, warnOnError } =
+    config.build.dynamicImportVarsOptions
+  const filter = createFilter(include, exclude)
+  const isBuild = config.command === 'build'
+  return {
+    name: 'vite:dynamic-import-vars',
+
+    resolveId(id) {
+      if (id === dynamicImportHelperId) {
+        return id
+      }
+    },
+
+    load(id) {
+      if (id === dynamicImportHelperId) {
+        return 'export default ' + dynamicImportHelper.toString()
+      }
+    },
+
+    async transform(source, importer) {
+      if (!filter(importer)) {
+        return
+      }
+
+      await init
+
+      let imports: readonly ImportSpecifier[] = []
+      try {
+        imports = parseImports(source)[0]
+      } catch (e: any) {
+        // ignore as it might not be a JS file, the subsequent plugins will catch the error
+        return null
+      }
+
+      if (!imports.length) {
+        return null
+      }
+
+      let s: MagicString | undefined
+      let needDynamicImportHelper = false
+
+      for (let index = 0; index < imports.length; index++) {
+        const {
+          s: start,
+          e: end,
+          ss: expStart,
+          se: expEnd,
+          d: dynamicIndex
+        } = imports[index]
+
+        if (dynamicIndex === -1 || source[start] !== '`') {
+          continue
+        }
+
+        s ||= new MagicString(source)
+        let result
+        try {
+          result = await transformDynamicImport(
+            source.slice(start, end),
+            importer,
+            resolve
+          )
+        } catch (error) {
+          if (warnOnError) {
+            this.warn(error)
+          } else {
+            this.error(error)
+          }
+        }
+
+        if (!result) {
+          continue
+        }
+
+        const { rawPattern, glob } = result
+
+        needDynamicImportHelper = true
+        s.overwrite(
+          expStart,
+          expEnd,
+          `__variableDynamicImportRuntimeHelper(${glob}, \`${rawPattern}\`)`
+        )
+      }
+
+      if (s) {
+        if (needDynamicImportHelper) {
+          s.prepend(
+            `import __variableDynamicImportRuntimeHelper from "${dynamicImportHelperId}";`
+          )
+        }
+        return {
+          code: s.toString(),
+          map: isBuild
+            ? config.build.sourcemap
+              ? s.generateMap({ hires: true })
+              : null
+            : s.generateMap({ hires: true })
```

Comment:
```suggestion
          map: !isBuild || config.build.sourcemap
            ? s.generateMap({ hires: true })
            : null
```

I think this can be simplified.